pkg/component/controller/extensions_controller.go

```diff
@@ -515,7 +522,13 @@ func (ec *ExtensionsController) startControllerManager(ctx context.Context) {
 // Stop
 func (ec *ExtensionsController) Stop() error {
 	ec.L.Info("Stopping extensions controller")
-	ec.mgrCancelFn()
+	// We have no guarantees on concurrency here, so use mutex
+	ec.mux.Lock()
```

Comment:
> is not nil, changes to a different cancel function for whatever reason and we call the old cancel function.

That's the culprit IMO. In that case, the new context might not get cancelled. I agree we can get away without a mutex, if we prefer, but then we'd need atomics to swap values and cancel the swapped-out values (if not nil).

Reply:
> Correct me if I'm wrong, but if we're doing mgrCancelFn := ec.mgrCancelFn the lock serves no purpose at all.

It does: It prevents data races when reading `ec.mgrCancelFn`, i.e. it ensures that we see the current cancel function, not some other value that was _just_ overwritten concurrently. It's not necessary to hold the lock for longer than the variable read, as the actual cancel function implementations are concurrency-safe and we don't need to block until they returned. However, holding the lock until they returned doesn't harm either here.

Reply:
> It does: It prevents data races when reading `ec.mgrCancelFn`, i.e. it ensures that we see the current cancel function, not some other value that was just overwritten concurrently.

I'm not very convinced... If this goroutine that sets`mgrCancelFn := ec.mgrCancelFn` acquires the lock BEFORE the goroutine that sets a new `ec.mgrCancelFn` you have effectively the same problem and I don't see how can a lock help here...

The only thing that would prevent this would be:

```
	ec.mux.Lock()
	if ec.mgrCancelFn != nil {
		ec.mgrCancelFn()
	}
	ec.mux.Unlock()
```

However at the end of the day it doesn't do much either because you don't have any guarantees that when a function checks if the context is done the new context may have been overwritten.

Anyway it doesn't do any harm so I guess we can ignore this for now and worry about these details in https://github.com/k0sproject/k0s/pull/4733 and subsequent work...

Reply:
> I'm not very convinced... If this goroutine that setsmgrCancelFn := ec.mgrCancelFn acquires the lock BEFORE the goroutine that sets a new ec.mgrCancelFn you have effectively the same problem and I don't see how can a lock help here...

That's right. The lock here only prevents the data race. The places where the cancel func gets written also need to ensure that the previous cancel func gets called.

I don't see how this could be fixed by holding the lock longer here, though.

Reply:
> The places where the cancel func gets written also need to ensure that the previous cancel func gets called.

Added this